### ISSUE TITLE
Question: Do you plan to support Framework 7 version 2?

### ISSUE THREAD

**Issue by rossmartin** [bencompton/framework7-react#77]
@bencompton @jascolley 

Thank you for this bringing this great framework to react.  I am considering using this in a production app and I'm curious if you plan to maintain this and keep it in sync with the production releases of Framework 7.  Do either of you feel that F7 react is ready for prime time?  I have a project I've started but I'm not too far in and I could change to react native before I get too much further.  I'm concerned about finding a serious issue 6 months from now and then not having any support.

**Comment by bencompton** [comment]
@jascolley and I use this library in a production app ourselves that we are actively developing and expect to be around for years to come. We do our best to keep up with new versions of framework7-vue, but maintaining this library isn't a full-time effort, so you might have to wait a while before updates are released or bugs are are addressed. Framework7 v2 is definitely in the roadmap, though.

If you want the bleeding edge of Framework7 components as timely as possible, Framework7 Vue is the upstream library that powers this one, and Vue is a nice alternative to React. All of the components in this library are auto-generated from framework7-vue, though, so the two libraries often have the same bugs.

React Native has a much larger community and a huge company backing it, and so support will likely be better than Framework7, but you will be giving up the advantages of a web app of course. You might consider building a small prototype with each option and see which is a better fit for your team and project. If you decide to use framework7-react and we don't fix your issues fast enough, we are always open to PRs. ;)

**Comment by rossmartin** [comment]
@bencompton Thanks for providing all the info and explaining.  It is a relief to hear that you plan to maintain this project.  If I find an opportunity to contribute I certainly will.

I've been using Cordova since 2012 and worked on several production apps using it and 2 personal apps released under my LLC.  I just developed my first production app with react native this year and released it in August (https://itunes.apple.com/us/app/hey-camera-voice-controlled-camera/id1271273970?mt=8).  My experience with react native was great but I would prefer a hybrid app based on the requirements and needs of this new app for my day job.  I used Ionic 2 for another personal app (https://itunes.apple.com/us/app/pet-search-adopt-a-pet-near-you/id1127973266?mt=8) and I cringe thinking about dealing with the new angular again (whatever version # they call it now lol).  

I've found this to be the best react based UI framework for a cordova app and thank you again for making that happen.

**Comment by rossmartin** [comment]
@bencompton It appears that v2 stable has been released for [framework7-vue](https://github.com/framework7io/framework7-vue) but the docs have not been updated yet on the site.  Do you have a time estimate on when you plan to get framework7-react updated?

**Comment by rossmartin** [comment]
@bencompton I came to the repo today to see if there were any updates and saw you've been working in a v2 branch getting things updated.  Thanks for continuing to support the project.  

I want to let you know that I ended up moving the project I'm working on to f7 v2 and scrapped f7-react because we weren't that far along and I kept finding issues.  I've had good luck integrating react 16 with f7 v2 and got the default f7 router to render react components as pages (and also got the f7 router and f7 instance passed as props to the react components). 

Before I elaborate further I want to say I appreciate your work and like what you've done with f7-react.  In my opinion I think f7 would be best placed as framework agnostic and not have vue/react bindings.  Here are a couple issues filed on the official f7 repo that have several valid points 

[allow "router-component" to be written in jsx](https://github.com/framework7io/framework7/issues/2091)
[use "vue jsx" to provide easier migration to react or jump to react](https://github.com/framework7io/framework7/issues/2055)

After reading the f7 v2 docs on its router and thanks to @nolimits4web [answer in the above issue](https://github.com/framework7io/framework7/issues/2091#issuecomment-357180907) it was relatively easy to get the built in f7 router to work with react components.

The TLDR is that I think you should use your skills and knowledge to get f7 usable with react out of the box and forget about bindings and simplifying components.  Let me know your thoughts on this.  And again I appreciate your time and work on this.

@nimo23

**Comment by bencompton** [comment]
@rossmartin - Sounds like this path is definitely worth investigating. Thanks for the info--I'll have a look.

**Comment by bencompton** [comment]
@rossmartin - I reviewed the links you provided. It's really awesome that you were able to leverage the extensibility of rendering in V2 and get React integration working. I'd be interested in seeing some code samples and learning more about your experiences with that approach. If it is working well for you, then you might consider open sourcing a starter template.

In my mind, the main value that the Vue and React libraries provide is to save everyone from re-inventing the wheel. For example, before Framework7 React, my application components were rendering F7 HTML directly. Then I found that there was a lot of duplicate code to render the same HTML over and over again, and I also didn't like having Framework7 API calls scattered throughout the app. I wanted to be able to declaratively send props into my React components and do as little imperative logic as possible. So, the next step was to move the F7-specific logic into separate components, which eventually became the first iteration of Framework7 React. I can imagine that many people building a reasonably large application in React or Vue might go through a similar thought process.

Perhaps you could expand on why you think that the Vue and React bindings are not the best use of resources? Is your thought that many people simply want a way to use Vue and React with Framework7 and don't really care as much about having the rendering logic and Framework7 API encapsulation done for them? Alternatively, is it just that you found it easier to handle the rendering and F7 API calls yourself rather than work around the issues you were finding and lack of V2 support?

**Comment by rossmartin** [comment]
@bencompton thank you for the response and explanations.  

I agree that there are advantages to using react bindings rather than using raw HTML.  Overall simplicity and consistency are key which I agree can be best accomplished by having react bindings for the f7 components.  Sorry for saying to forget about vue/react bindings - they simplify things but consistency is important.  One problem I kept running into was that attributes didn't work on the react-f7 components.  It was either they didn't add the right class or output the correct HTML.  This resulted in code having a mixture of react-f7 components and raw HTML which started to get confusing.  It was hard to tell if the problem was in f7-vue or f7-react.

The documentation was an area that was troublesome for me also.  The f7-vue/f7-react docs are not as complete and the examples on components don't show what it looks like on an ios/android device within the page.  You'll find yourself referring to the vanilla f7 docs often and then get confused - something like "can I do this same thing but declaratively?".  I think you "reactified" the vue docs and I'm not trying to point a finger at anyone for incomplete docs.

One thing that isn't clear to me is what was done in the `framework7-bencompton-fork` - this is a dev dependency of this repo here and it isn't on GitHub and only published to npm.  This furthers my confusion of "is the bug in f7, f7-vue, or f7-react"?  I wondered "has the core framework 7 code been modified to make f7-react"?

At f7-react's current stage I have found it easier to handle the rendering and f7 API calls myself rather than working around the issues I've come across.  Development is ramping up on this project of ours and I didn't want to keep moving forward with the current f7-react and then have to deal with upgrading down the road - it wasn't clear when this would be updated to v2.  So far I've had good luck at using v2 with react.  Below is a simplified snippet of how I'm dealing with routing - 

`routes.js`
```javascript
import React from 'react';
import ReactDOM from 'react-dom';

import App from './components/App';
import SomePage from './components/pages/SomePage';

export default [
  {
    path: '/',
    async: (routeTo, routeFrom, resolve, reject) => {
      return resolve({
        component: reactPageComponentWrapper(SomePage)
      });
    }
  }
];

function reactPageComponentWrapper(ComponentRef) {
  return {
    render() {
      return '<div class="page"></div>';
    },
    on: {
      pageBeforeIn(event, page) {
        const { f7, currentUser } = App;
        const router = f7.views.main.router;

        ReactDOM.render(
          <ComponentRef f7={f7} router={router} />,
          page.$el[0]
        );
      }
    }
  };
}
```

Animations are working using this approach and I believe the page/component lifecycle is being handled correctly.  The `componentWillReceiveProps` events fires when going back to a rendered react component page (allowing you to check props and deal with a "page:reinit").

**Comment by rossmartin** [comment]
@bencompton @nimo23 I want to share a better way of using the f7 router that I have discovered.  This might help someone in the future.  The flaw with the above approach is that some f7 components (particularly swipeable tabs) rely on content being in the DOM immediately after the route change in order to bind events for the f7 component.

This is a sound approach that I have found - 

```javascript
import React from 'react';
import ReactDOM from 'react-dom';

import App from './components/App';
import SomePage from './components/pages/SomePage';

export default [
  {
    path: '/',
    async: renderReactComponent.bind(this, SomePage)
  }
];

function renderReactComponent(ComponentRef, routeTo, routeFrom, resolve, reject) {
  const container = document.createElement('div');

  ReactDOM.render(
    <ComponentRef />
    container,
    () => {
      resolve({ el: container });
    }
  );
}
```

**Comment by bencompton** [comment]
@rossmartin - Awesome work there--thanks for sharing!

**Comment by DuffmanBE** [comment]
@rossmartin - Because of performance problems we are experiencing due to the router of the framework7-react package, we are planning on switching to the React-router. We are however experiencing some problems. Would you be open to respond to a couple of questions we have?

**Comment by rossmartin** [comment]
@DuffmanBE Sure.  I've never used the react router with f7.  Like mentioned above I ended up using f7 v2 and making the f7 v2 router render react components.

You can contact me via email - ross at chattur.co
